usb/esp_tinyusb/tusb_msc_storage.c

```diff
@@ -0,0 +1,563 @@
+        data_partition = esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_DATA_FAT, NULL);
+        ESP_RETURN_ON_FALSE(data_partition, ESP_ERR_NOT_FOUND, TAG,
+                            "Failed to find FATFS partition. Check the partition table.");
+    }
+
+    return wl_mount(data_partition, &s_storage_handle->wl_handle);
+}
+
+static esp_err_t _mount_spiflash(BYTE pdrv)
+{
+    return ff_diskio_register_wl_partition(pdrv, s_storage_handle->wl_handle);
+}
+
+static esp_err_t _unmount_spiflash(void)
+{
+    BYTE pdrv;
+    pdrv = ff_diskio_get_pdrv_wl(s_storage_handle->wl_handle);
+    if (pdrv == 0xff) {
+        ESP_LOGE(TAG, "Invalid state");
+        return ESP_ERR_INVALID_STATE;
+    }
+    ff_diskio_clear_pdrv_wl(s_storage_handle->wl_handle);
+
+    char drv[3] = {(char)('0' + pdrv), ':', 0};
+    f_mount(0, drv, 0);
+    ff_diskio_unregister(pdrv);
+
+    return ESP_OK;
+}
+
+static uint32_t _get_sector_count_spiflash(void)
+{
+    uint32_t result = 0;
+    assert(s_storage_handle->wl_handle != WL_INVALID_HANDLE);
+    size_t size = wl_sector_size(s_storage_handle->wl_handle);
+    if (size == 0) {
+        ESP_LOGW(TAG, "WL Sector size is zero !!!");
+        result = 0;
+    } else {
+        result = (uint32_t)(wl_size(s_storage_handle->wl_handle) / size);
+    }
+    return result;
+}
+
+static uint32_t _get_sector_size_spiflash(void)
+{
+    assert(s_storage_handle->wl_handle != WL_INVALID_HANDLE);
+    return (uint32_t)wl_sector_size(s_storage_handle->wl_handle);
+}
+
+static esp_err_t _read_sector_spiflash(size_t sector_size,
+                                       uint32_t lba,
+                                       uint32_t offset,
+                                       size_t size,
+                                       void *dest)
+{
+    size_t addr = lba * sector_size + offset; // Address of the data to be read, relative to the beginning of the partition.
+    return wl_read(s_storage_handle->wl_handle, addr, dest, size);
+}
+
+static esp_err_t _write_sector_spiflash(size_t sector_size,
+                                        size_t addr,
+                                        uint32_t lba,
+                                        uint32_t offset,
+                                        size_t size,
+                                        const void *src)
+{
+    ESP_RETURN_ON_ERROR(wl_erase_range(s_storage_handle->wl_handle, addr, size),
+                        TAG, "Failed to erase");
+    return wl_write(s_storage_handle->wl_handle, addr, src, size);
+}
+
+#if SOC_SDMMC_HOST_SUPPORTED
+static esp_err_t _init_sdmmc(const tinyusb_msc_storage_config_t *config)
+{
+    esp_err_t ret = ESP_OK;
+
+    sdmmc_host_t host = config->sdmmc_config.host;
+    sdmmc_slot_config_t slot_config = config->sdmmc_config.slot_config;
+
+    ESP_RETURN_ON_ERROR((*host.init)(), TAG, "Host Config Init fail");
+
+    ESP_GOTO_ON_ERROR(sdmmc_host_init_slot(host.slot, (const sdmmc_slot_config_t *) &slot_config),
```

Comment:
What about keeping the sdmmc initialization outside of this component, in the user's code?

The advantage of such approach is that if the user's application has initialized the card, you can access both cards over SDMMC and SDSPI using the same API.

It is also hard to explain to the user why the SD card initialization in their application would be handled by... esp_tinyusb component, of all the things.

How about changing the initialization APIs to:

```c++
// sdmmc_card_t* card is now part of "tinyusb_msc_sdmmc_config_t"
tinyusb_msc_storage_init_sdmmc(const tinyusb_msc_sdmmc_config_t *config); 

tinyusb_msc_storage_init_spiflash(const tinyusb_msc_spiflash_config_t* config);
```

Both functions can initialize `s_storage_handle`, as your `register...` APIs do now, and then call the common internal `init` implementation for both cases.

Reply:
The initialisation part is now shifted to user application for both spiflash and sdmmc to keep the symmetry.
APIs are changed accordingly.
@igrr   PTAL.